### PR TITLE
[JUJU-3935] Update apiserver resources to accept worker.Worker

### PR DESCRIPTION
**Pull request by nvinuesa**
API resources for all intents and purposes are `worker.Worker` types. The API resource package predates catacombs and the v2 tomb package, so is legacy code that could and should be bound to a catacomb. The following lays out the land to move all watchers to `worker.Worker`. This includes replacing the `Stop` with what most do internally; `Kill` and `Wait`.

The additional change is move Pinger (previously PingerTimeout) to a real facade. This then allows the changing the Pinger to become a real `worker.Worker` type without exposing the `Kill` and `Wait` to the API. This removes the old TODO from @SimonRichardson which effectively states that the Pinger type shouldn't be exposed directly.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Just verify it builds and tests pass.

```
make go-build
```

### CODE REVIEW

File: apiserver/facades/controller/charmdownloader/interfaces.go [hunk 1]
```diff
@@ -63,5 +64,5 @@ type ResourcesBackend interface {
 
 // StoppableResource is implemented by resources that can be stopped.
 type StoppableResource interface {
-	Stop() error
+	worker.Worker
 }
```
Comment (SimonRichardson) [inline apiserver/facades/controller/charmdownloader/interfaces.go:68]:
```
type StoppableResource = worker.Worker
```

File: apiserver/facades/schema.json [hunk 2]
```diff
@@ -33683,11 +33683,16 @@
         "Schema": {
             "type": "object",
             "properties": {
+                "Kill": {
```
Comment (SimonRichardson) [inline apiserver/facades/schema.json:33686]: I wonder if we should still keep `Stop` for Pinger to keep backwards compatibility.

File: apiserver/facades/controller/charmdownloader/interfaces.go (hunk 1 above)
Comment (nvinuesa) [reply apiserver/facades/controller/charmdownloader/interfaces.go]: done

File: apiserver/facades/schema.json (hunk 2 above)
Comment (nvinuesa) [reply apiserver/facades/schema.json:33686]: done in 2749e98dae

**Review by SimonRichardson — CHANGES_REQUESTED** [review]
The `statetesting.AssertKillAndWait` should actually be `workertest.CleanKill` and any `defer statetesting.AssertKillAndWait` should be `defer workertest.DirtyKill`

**Review by SimonRichardson — COMMENTED** [review]
Looks good, let's get @manadart to approve it.

**Review by SimonRichardson — COMMENTED** [review]
Looks good, let's get @manadart to approve it.

File: apiserver/common/resource.go [hunk 3]
```diff
@@ -129,10 +131,19 @@ func (rs *Resources) Count() int {
...
 type StringResource string
 
+func (StringResource) Kill() {
+}
+
+func (StringResource) Wait() error {
+	return nil
+}
+
 func (StringResource) Stop() error {
```
Comment (manadart) [inline apiserver/common/resource.go:147]: I know these are temporary, but technically we can remove the `Stop` methods here right?

File: core/multiwatcher/types.go [hunk 4]
```diff
@@ -41,7 +42,8 @@ type Factory interface {
 // Watcher is the way a caller can find out what changes have happened
 // on one or more models.
 type Watcher interface {
-	Stop() error
+	worker.Worker
+	// Stop() error
```
Comment (manadart) [inline core/multiwatcher/types.go:46]: Remove this.

**Review by manadart — COMMENTED** [review]
Almost there.

File: state/state_test.go [hunk 5]
```diff
@@ -2662,7 +2663,7 @@ func (s *StateSuite) TestWatchModelsBulkEvents(c *gc.C) {
 
 	// All except the removed model are reported in initial event.
 	w := s.State.WatchModels()
-	defer statetesting.AssertStop(c, w)
+	defer workertest.DirtyKill(c, w)
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply state/state_test.go:2666]: I followed https://github.com/juju/juju/pull/15705#pullrequestreview-1469281585, should I change all `defer workertest.DirtyKill` to `defer workertest.CleanKill` everywhere or just here?

File: core/multiwatcher/types.go (hunk 4 above)
Comment (nvinuesa) [reply core/multiwatcher/types.go]: done

File: apiserver/common/resource.go (hunk 3 above)
Comment (nvinuesa) [reply apiserver/common/resource.go]: sure, removed

File: state/state_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nvinuesa) [reply state/state_test.go]: done

**Review by SimonRichardson — APPROVED** [review]

**Review by manadart — APPROVED** [review]

**Pull request merged**